iroha/src/sumeragi.rs

```diff
@@ -656,343 +659,561 @@ impl<Q: QueueTrait, G: GenesisNetworkTrait, W: WorldTrait> Sumeragi<Q, G, W> {
+
+    use iroha_error::{error, Result};
+    use parity_scale_codec::{Decode, Encode};
+    use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};
+
+    use crate::block::EmptyChainHash;
+
+    /// Sorts peers based on the `hash`.
+    pub fn sort_peers_by_hash(peers: Vec<PeerId>, hash: Hash) -> Vec<PeerId> {
+        sort_peers_by_hash_and_counter(peers, hash, 0)
+    }
+
+    /// Sorts peers based on the `hash` and `counter` combined as a seed.
+    pub fn sort_peers_by_hash_and_counter(
+        mut peers: Vec<PeerId>,
+        hash: Hash,
+        counter: u32,
+    ) -> Vec<PeerId> {
+        peers.sort_by(|p1, p2| p1.address.cmp(&p2.address));
+        let mut bytes: Vec<u8> = counter.to_le_bytes().to_vec();
+        bytes.append(hash.as_ref().to_vec().as_mut());
+        let Hash(bytes) = Hash::new(&bytes);
+        let mut rng = StdRng::from_seed(bytes);
+        peers.shuffle(&mut rng);
+        peers
+    }
+
+    /// Shifts `sorted_peers` by one to the right.
+    #[allow(clippy::expect_used)]
+    pub fn shift_peers_by_one(mut peers: Vec<PeerId>) -> Vec<PeerId> {
+        let last_element = peers.pop().expect("No elements found in sorted peers.");
+        peers.insert(0, last_element);
+        peers
+    }
+
+    /// Shifts `sorted_peers` by `n` to the right.
+    pub fn shift_peers_by_n(mut peers: Vec<PeerId>, n: u32) -> Vec<PeerId> {
+        for _ in 0..n {
+            peers = shift_peers_by_one(peers);
         }
+        peers
     }
 
-    /// Initializes network topology.
-    ///
-    /// # Errors
-    /// Can fail if consensus criteria is not meet:
-    /// `len(peers) >= 3 * max_faulty_peers + 1`
-    pub fn init(self) -> Result<InitializedNetworkTopology> {
-        let min_peers = 3 * self.max_faults + 1;
-        if self.peers.len() >= min_peers as usize {
-            let mut topology = InitializedNetworkTopology {
-                sorted_peers: self.peers.into_iter().collect(),
-                max_faults: self.max_faults,
-            };
-            topology.sort_peers_by_hash(self.block_hash);
-            Ok(topology)
-        } else {
-            Err(error!(
-                "Not enough peers to be Byzantine fault tolerant. Expected a least {} peers, got {}",
-                3 * self.max_faults + 1,
-                self.peers.len(),
+    macro_rules! field_is_some_or_err {
+        ($s:ident.$f:ident) => {
+            $s.$f.ok_or(error!(
+                "Field with name {} should not be `None`.",
+                stringify!($f)
             ))
-        }
+        };
     }
-}
 
-/// Represents a topology of peers, defining a `role` for each peer based on the previous block hash.
-#[derive(Debug, Clone, Encode, Decode)]
-pub struct InitializedNetworkTopology {
-    /// Current order of peers. The roles of peers are defined based on this order.
-    pub sorted_peers: Vec<PeerId>,
-    /// Maximum faulty peers in a network.
-    pub max_faults: u32,
-}
+    pub struct GenesisBuilder {
```

Comment:
Should we just use some derive macro that implements builder pattern for structures?